File: adapters/mjx/scale.py
"""Measure selected accelerator batch widths without timing simulation logic."""

from __future__ import annotations

import argparse
import hashlib
import json
import math
import os
import time
from copy import deepcopy
from pathlib import Path
from typing import Any

from rne_mjx_adapter.backend import TASK_ID
from rne_mjx_adapter.protocol import (
    ADAPTER_ID,
    SCALE_REPORT_KIND,
    SCALE_REPORT_SCHEMA_VERSION,
    SUPPORTED_BATCH_WIDTHS,
    ProtocolError,
    canonical_json,
    load_json_fixture,
)
from rne_mjx_adapter.server import create_backend

DEFAULT_ROOT_SEED = 42
DEFAULT_WARMUP_STEPS = 32
DEFAULT_MEASURED_STEPS = 256


def build_scale_report(
    adapter_root: Path,
    *,
    backend_name: str,
    allow_test_backend: bool,
    widths: list[int],
    warmup_steps: int,
    measured_steps: int,
) -> dict[str, Any]:
    """Measures fixed-step batches and proves width-independent lane zero."""

    _validate_positive_steps(warmup_steps, "warmup_steps", allow_zero=True)
    _validate_positive_steps(measured_steps, "measured_steps", allow_zero=False)
    if not widths or widths != sorted(set(widths)):
        raise ProtocolError(
            "invalid_widths",
            "widths must be a non-empty, strictly increasing unique list",
        )
    if any(width not in SUPPORTED_BATCH_WIDTHS for width in widths):
        raise ProtocolError(
            "unsupported_batch_width",
            f"widths must be selected from {list(SUPPORTED_BATCH_WIDTHS)}",
        )

    fixtures = adapter_root / "fixtures"
    task_spec = load_json_fixture(fixtures / "free-fall-task-spec-v1.json")
    model_text = (fixtures / "free-fall-v1.xml").read_text(encoding="utf-8")
    backend = create_backend(backend_name, adapter_root, allow_test_backend)
    capability = backend.capability_report()
    if capability["status"] == "unavailable":
        raise ProtocolError(
            "runtime_unavailable",
            "accelerator scale runtime is unavailable",
            details={"reason_code": capability["unavailable_reason_code"]},
        )

    runs: list[dict[str, Any]] = []
    lane_zero_digests: list[int] = []
    for width in widths:
        session = backend.create_session(
            task_spec,
            DEFAULT_ROOT_SEED,
            width,
            True,
        )
        actions = [[0.0] for _ in range(width)]
        result: dict[str, Any] | None = None
        for _ in range(warmup_steps):
            result = session.step(actions)
        started_ns = time.perf_counter_ns()
        for _ in range(measured_steps):
            result = session.step(actions)
        elapsed_ns = max(1, time.perf_counter_ns() - started_ns)
        assert result is not None
        transitions = width * measured_steps
        throughput_transitions_s = transitions * 1_000_000_000.0 / elapsed_ns
        if not math.isfinite(throughput_transitions_s):
            raise ProtocolError("non_finite_timing", "scale throughput is not finite")
        lane_zero_digest = result["lane_replay_digests"][0]
        lane_zero_digests.append(lane_zero_digest)
        runs.append(
            {
                "batch_width": width,
                "transitions": transitions,
                "elapsed_ns": elapsed_ns,
                "throughput_transitions_s": throughput_transitions_s,
                "lane_zero_replay_digest": lane_zero_digest,
                "lane_zero_episode_index": result["episode_indices"][0],
                "lane_zero_episode_seed": result["episode_seeds"][0],
            }
        )

    lane_zero_digest_consistent = len(set(lane_zero_digests)) == 1
    promotion_widths_complete = widths == list(SUPPORTED_BATCH_WIDTHS)
    evidence_class = (
        "contract_test" if capability["status"] == "test_only" else "accelerator"
    )
    passed = lane_zero_digest_consistent and (
        evidence_class == "contract_test" or promotion_widths_complete
    )
    report: dict[str, Any] = {
        "kind": SCALE_REPORT_KIND,
        "schema_version": SCALE_REPORT_SCHEMA_VERSION,
        "adapter_id": ADAPTER_ID,
        "evidence_class": evidence_class,
        "backend_status": capability["status"],
        "precision": capability["precision"],
        "measurement_boundary": "python_session_api",
        "task_id": TASK_ID,
        "task_spec_schema": task_spec["schema_version"],
        "task_spec_sha256": _sha256_text(canonical_json(task_spec)),
        "model_sha256": _sha256_text(model_text.replace("\r\n", "\n")),
        "root_seed": DEFAULT_ROOT_SEED,
        "warmup_steps": warmup_steps,
        "measured_steps": measured_steps,
        "requested_widths": list(widths),
        "promotion_widths_complete": promotion_widths_complete,
        "lane_zero_digest_consistent": lane_zero_digest_consistent,
        "runs": runs,
        "runtime_contract": deepcopy(capability["runtime_contract"]),
        "runtime": deepcopy(capability["runtime"]),
        "passed": passed,
    }
    report["content_sha256"] = _report_digest(report)
    validate_scale_report(report)
    return report


def validate_scale_report(report: Any) -> dict[str, Any]:
    """Validates the scale-report envelope, finite timings, and digest."""

    if not isinstance(report, dict):
        raise ProtocolError("report_invalid", "scale report must be an object")
    required = {
        "kind",
        "schema_version",
        "adapter_id",
        "evidence_class",
        "backend_status",
        "precision",
        "measurement_boundary",
        "task_id",
        "task_spec_schema",
        "task_spec_sha256",
        "model_sha256",
        "root_seed",
        "warmup_steps",
        "measured_steps",
        "requested_widths",
        "promotion_widths_complete",
        "lane_zero_digest_consistent",
        "runs",
        "runtime_contract",
        "runtime",
        "passed",
        "content_sha256",
    }
    if set(report) != required:
        raise ProtocolError("report_invalid", "scale report fields do not match schema")
    if report["kind"] != SCALE_REPORT_KIND or report["schema_version"] != SCALE_REPORT_SCHEMA_VERSION:
        raise ProtocolError("report_invalid", "scale report kind or schema mismatch")
    if (
        report["adapter_id"] != ADAPTER_ID
        or report["task_id"] != TASK_ID
        or report["task_spec_schema"] != 1
        or report["precision"] != "f64"
    ):
        raise ProtocolError("report_invalid", "scale contract identity mismatch")
    if report["evidence_class"] not in {"contract_test", "accelerator"}:
        raise ProtocolError("report_invalid", "unknown scale evidence class")
    if report["evidence_class"] == "contract_test" and report["backend_status"] != "test_only":
        raise ProtocolError("report_invalid", "contract-test scale must use test_only backend")
    if report["measurement_boundary"] != "python_session_api":
        raise ProtocolError("report_invalid", "scale measurement boundary mismatch")
    if not isinstance(report["runs"], list) or not report["runs"]:
        raise ProtocolError("report_invalid", "scale report requires at least one run")
    if (
        not isinstance(report["requested_widths"], list)
        or len(report["runs"]) != len(report["requested_widths"])
        or report["requested_widths"] != sorted(set(report["requested_widths"]))
        or any(width not in SUPPORTED_BATCH_WIDTHS for width in report["requested_widths"])
    ):
        raise ProtocolError("report_invalid", "scale run count differs from requested widths")
    _validate_positive_steps(report["warmup_steps"], "warmup_steps", allow_zero=True)
    _validate_positive_steps(report["measured_steps"], "measured_steps", allow_zero=False)
    run_fields = {
        "batch_width",
        "transitions",
        "elapsed_ns",
        "throughput_transitions_s",
        "lane_zero_replay_digest",
        "lane_zero_episode_index",
        "lane_zero_episode_seed",
    }
    for index, run in enumerate(report["runs"]):
        if (
            not isinstance(run, dict)
            or set(run) != run_fields
            or run.get("elapsed_ns", 0) <= 0
            or not math.isfinite(run.get("throughput_transitions_s", float("nan")))
            or run.get("transitions") != run.get("batch_width", 0) * report["measured_steps"]
        ):
            raise ProtocolError("report_invalid", "scale run timing is invalid")
        if run["batch_width"] != report["requested_widths"][index]:
            raise ProtocolError("report_invalid", "scale run order differs from requested widths")
        expected_throughput = run["transitions"] * 1_000_000_000.0 / run["elapsed_ns"]
        if run["throughput_transitions_s"] != expected_throughput:
            raise ProtocolError("report_invalid", "scale throughput was not recomputed")
    digest_consistent = len(
        {run["lane_zero_replay_digest"] for run in report["runs"]}
    ) == 1
    if report["lane_zero_digest_consistent"] != digest_consistent:
        raise ProtocolError("report_invalid", "lane-zero consistency verdict mismatch")
    promotion_complete = report["requested_widths"] == list(SUPPORTED_BATCH_WIDTHS)
    if report["promotion_widths_complete"] != promotion_complete:
        raise ProtocolError("report_invalid", "promotion-width verdict mismatch")
    expected_passed = digest_consistent and (
        report["evidence_class"] == "contract_test" or promotion_complete
    )
    if report["passed"] != expected_passed:
        raise ProtocolError("report_invalid", "scale pass verdict mismatch")
    canonical_json(report)
    if report["content_sha256"] != _report_digest(report):
        raise ProtocolError("report_digest_mismatch", "scale report digest mismatch")
    return report


def write_scale_report(path: Path, report: dict[str, Any]) -> None:
    """Atomically writes a validated scale report."""

    validate_scale_report(report)
    path.parent.mkdir(parents=True, exist_ok=True)
    payload = json.dumps(report, allow_nan=False, indent=2, sort_keys=True) + "\n"
    for attempt in range(100):
        temporary = path.with_name(f".{path.name}.tmp-{attempt}")
        try:
            destination = temporary.open("x", encoding="utf-8", newline="\n")
        except FileExistsError:
            continue
        try:
            with destination:
                destination.write(payload)
                destination.flush()
                os.fsync(destination.fileno())
            os.replace(temporary, path)
            return
        except Exception:
            temporary.unlink(missing_ok=True)
            raise
    raise ProtocolError(
        "temporary_file_exhausted",
        "could not reserve a bounded scale-report temporary file",
    )


def main(argv: list[str] | None = None) -> int:
    """Runs all selected widths and optionally writes the report."""

    parser = argparse.ArgumentParser(description="RNE MJX-Warp scale report")
    parser.add_argument("--backend", choices=("mjx_warp", "fake"), default="mjx_warp")
    parser.add_argument("--allow-test-backend", action="store_true")
    parser.add_argument("--widths", default=",".join(str(width) for width in SUPPORTED_BATCH_WIDTHS))
    parser.add_argument("--warmup-steps", type=int, default=DEFAULT_WARMUP_STEPS)
    parser.add_argument("--measured-steps", type=int, default=DEFAULT_MEASURED_STEPS)
    parser.add_argument("--output", type=Path)
    args = parser.parse_args(argv)
    try:
        widths = [int(value) for value in args.widths.split(",")]
        report = build_scale_report(
            Path(__file__).resolve().parent,
            backend_name=args.backend,
            allow_test_backend=args.allow_test_backend,
            widths=widths,
            warmup_steps=args.warmup_steps,
            measured_steps=args.measured_steps,
        )
        if args.output is not None:
            write_scale_report(args.output, report)
        print(canonical_json(report))
        return 0 if report["passed"] else 1
    except (OSError, ProtocolError, ValueError) as error:
        if isinstance(error, ProtocolError):
            payload = error.as_object()
        else:
            payload = {
                "code": "invalid_scale_request",
                "message": "scale request or I/O failed",
                "details": {},
            }
        print(canonical_json({"kind": "rne_accelerator_scale_error", "error": payload}))
        return 2


def _validate_positive_steps(value: int, field: str, *, allow_zero: bool) -> None:
    minimum = 0 if allow_zero else 1
    if isinstance(value, bool) or not isinstance(value, int) or not minimum <= value <= 1_000_000:
        raise ProtocolError(
            "invalid_steps",
            f"{field} must be an integer in {minimum}..=1000000",
        )


def _report_digest(report: dict[str, Any]) -> str:
    without_digest = {key: value for key, value in report.items() if key != "content_sha256"}
    return _sha256_text(canonical_json(without_digest))


def _sha256_text(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/rne_accelerator_contract/src/scale.rs
//! Accelerator scale-report v1 reader.

use super::conformance::{
    canonical_python_json, normalized_model_sha256, sha256, task_spec_sha256,
    validate_evidence_runtime, validate_hex_sha256,
};
use super::{
    invalid, require, require_identifier, AcceleratorContractError, AcceleratorManifest,
    AcceleratorRuntimeContract, AcceleratorRuntimeProbe, ACCELERATOR_SCALE_REPORT_SCHEMA_VERSION,
};
use rne_ai::{derive_episode_seed, TaskSpec, TASK_SPEC_SCHEMA_VERSION};
use serde::{Deserialize, Serialize};
use serde_json::value::RawValue;

/// Stable scale-report discriminator.
pub const ACCELERATOR_SCALE_REPORT_KIND: &str = "rne_accelerator_scale_report";

const DEFAULT_ROOT_SEED: u64 = 42;
const MAX_STEPS: u64 = 1_000_000;
const MAX_REPORT_BYTES: usize = 1024 * 1024;
const SUPPORTED_BATCH_WIDTHS: [usize; 4] = [1, 16, 256, 4096];

/// One measured batch width in scale-report v1.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AcceleratorScaleRun {
    /// Executed batch width.
    pub batch_width: usize,
    /// Exact `batch_width * measured_steps` transition count.
    pub transitions: u64,
    /// Measured monotonic duration in nanoseconds.
    pub elapsed_ns: u64,
    /// Recomputed transitions per second.
    pub throughput_transitions_s: f64,
    /// Same-build diagnostic lane-zero replay digest.
    pub lane_zero_replay_digest: u64,
    /// Lane-zero episode index after the measured steps.
    pub lane_zero_episode_index: u64,
    /// Deterministically derived seed for that episode index.
    pub lane_zero_episode_seed: u64,
}

/// Versioned batch-width and lane-isolation evidence for one accelerator.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AcceleratorScaleReport {
    /// Stable report discriminator.
    pub kind: String,
    /// Scale-report schema version.
    pub schema_version: u32,
    /// Stable adapter identifier.
    pub adapter_id: String,
    /// `contract_test` or physical `accelerator` evidence class.
    pub evidence_class: String,
    /// Capability status observed before execution.
    pub backend_status: String,
    /// Numeric precision used by every run.
    pub precision: String,
    /// Timed API boundary.
    pub measurement_boundary: String,
    /// Bound task identifier.
    pub task_id: String,
    /// Bound TaskSpec schema.
    pub task_spec_schema: u32,
    /// Lowercase SHA-256 of canonical TaskSpec JSON.
    pub task_spec_sha256: String,
    /// Lowercase SHA-256 of LF-normalized model text.
    pub model_sha256: String,
    /// Root seed shared by all widths.
    pub root_seed: u64,
    /// Untimed warm-up steps per width.
    pub warmup_steps: u64,
    /// Timed measured steps per width.
    pub measured_steps: u64,
    /// Strictly increasing requested widths.
    pub requested_widths: Vec<usize>,
    /// Whether all promotion widths were measured.
    pub promotion_widths_complete: bool,
    /// Whether lane-zero replay identity is width-independent.
    pub lane_zero_digest_consistent: bool,
    /// Ordered measured runs, one per requested width.
    pub runs: Vec<AcceleratorScaleRun>,
    /// Exact runtime requirements used during probing.
    pub runtime_contract: AcceleratorRuntimeContract,
    /// Runtime values observed during probing.
    pub runtime: AcceleratorRuntimeProbe,
    /// Verdict recomputed from evidence class, widths, and lane identity.
    pub passed: bool,
    /// Lowercase SHA-256 of Python-canonical report JSON excluding this field.
    pub content_sha256: String,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct RawScaleRun {
    batch_width: usize,
    transitions: u64,
    elapsed_ns: u64,
    throughput_transitions_s: Box<RawValue>,
    lane_zero_replay_digest: u64,
    lane_zero_episode_index: u64,
    lane_zero_episode_seed: u64,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct RawScaleReport {
    kind: String,
    schema_version: u32,
    adapter_id: String,
    evidence_class: String,
    backend_status: String,
    precision: String,
    measurement_boundary: String,
    task_id: String,
    task_spec_schema: u32,
    task_spec_sha256: String,
    model_sha256: String,
    root_seed: u64,
    warmup_steps: u64,
    measured_steps: u64,
    requested_widths: Vec<usize>,
    promotion_widths_complete: bool,
    lane_zero_digest_consistent: bool,
    runs: Vec<RawScaleRun>,
    runtime_contract: AcceleratorRuntimeContract,
    runtime: AcceleratorRuntimeProbe,
    passed: bool,
    content_sha256: String,
}

impl AcceleratorScaleReport {
    /// Parses bounded JSON while preserving throughput number lexemes.
    pub fn from_json_slice(bytes: &[u8]) -> Result<Self, AcceleratorContractError> {
        require(
            !bytes.is_empty() && bytes.len() <= MAX_REPORT_BYTES,
            "scale report JSON size is invalid",
        )?;
        let raw: RawScaleReport = serde_json::from_slice(bytes)
            .map_err(|error| invalid(format!("parse scale report JSON: {error}")))?;
        let runs = raw
            .runs
            .into_iter()
            .map(|run| {
                Ok(AcceleratorScaleRun {
                    batch_width: run.batch_width,
                    transitions: run.transitions,
                    elapsed_ns: run.elapsed_ns,
                    throughput_transitions_s: parse_raw_f64(&run.throughput_transitions_s)?,
                    lane_zero_replay_digest: run.lane_zero_replay_digest,
                    lane_zero_episode_index: run.lane_zero_episode_index,
                    lane_zero_episode_seed: run.lane_zero_episode_seed,
                })
            })
            .collect::<Result<Vec<_>, AcceleratorContractError>>()?;
        let report = Self {
            kind: raw.kind,
            schema_version: raw.schema_version,
            adapter_id: raw.adapter_id,
            evidence_class: raw.evidence_class,
            backend_status: raw.backend_status,
            precision: raw.precision,
            measurement_boundary: raw.measurement_boundary,
            task_id: raw.task_id,
            task_spec_schema: raw.task_spec_schema,
            task_spec_sha256: raw.task_spec_sha256,
            model_sha256: raw.model_sha256,
            root_seed: raw.root_seed,
            warmup_steps: raw.warmup_steps,
            measured_steps: raw.measured_steps,
            requested_widths: raw.requested_widths,
            promotion_widths_complete: raw.promotion_widths_complete,
            lane_zero_digest_consistent: raw.lane_zero_digest_consistent,
            runs,
            runtime_contract: raw.runtime_contract,
            runtime: raw.runtime,
            passed: raw.passed,
            content_sha256: raw.content_sha256,
        };
        report.validate()?;
        Ok(report)
    }

    /// Validates report-local timing, lane identity, promotion, runtime, and digest invariants.
    pub fn validate(&self) -> Result<(), AcceleratorContractError> {
        require(
            self.kind == ACCELERATOR_SCALE_REPORT_KIND,
            "scale-report kind mismatch",
        )?;
        require(
            self.schema_version == ACCELERATOR_SCALE_REPORT_SCHEMA_VERSION,
            "scale-report schema mismatch",
        )?;
        require_identifier(&self.adapter_id, "scale adapter id")?;
        require(self.precision == "f64", "scale precision must be f64")?;
        require(
            self.measurement_boundary == "python_session_api",
            "scale measurement boundary mismatch",
        )?;
        require_identifier(&self.task_id, "scale task id")?;
        require(
            self.task_spec_schema == TASK_SPEC_SCHEMA_VERSION,
            "scale TaskSpec schema mismatch",
        )?;
        validate_hex_sha256(&self.task_spec_sha256, "scale TaskSpec digest")?;
        validate_hex_sha256(&self.model_sha256, "scale model digest")?;
        validate_hex_sha256(&self.content_sha256, "scale content digest")?;
        require(
            self.root_seed == DEFAULT_ROOT_SEED,
            "scale root seed mismatch",
        )?;
        require(
            self.warmup_steps <= MAX_STEPS && (1..=MAX_STEPS).contains(&self.measured_steps),
            "scale step counts are out of bounds",
        )?;
        require(
            !self.requested_widths.is_empty()
                && self
                    .requested_widths
                    .windows(2)
                    .all(|pair| pair[0] < pair[1])
                && self
                    .requested_widths
                    .iter()
                    .all(|width| SUPPORTED_BATCH_WIDTHS.contains(width)),
            "scale requested widths are not canonical",
        )?;
        require(
            self.runs.len() == self.requested_widths.len(),
            "scale run count differs from requested widths",
        )?;
        for (run, requested_width) in self.runs.iter().zip(&self.requested_widths) {
            require(
                run.batch_width == *requested_width,
                "scale run order differs from requested widths",
            )?;
            let transitions = u64::try_from(run.batch_width)
                .ok()
                .and_then(|width| width.checked_mul(self.measured_steps))
                .ok_or_else(|| invalid("scale transition count overflowed"))?;
            require(
                run.transitions == transitions,
                "scale transition count was not recomputed",
            )?;
            require(run.elapsed_ns > 0, "scale elapsed time must be positive")?;
            let expected_throughput =
                run.transitions as f64 * 1_000_000_000.0 / run.elapsed_ns as f64;
            require(
                run.throughput_transitions_s.is_finite()
                    && run.throughput_transitions_s == expected_throughput,
                "scale throughput was not recomputed",
            )?;
            require(
                run.lane_zero_episode_seed
                    == derive_episode_seed(self.root_seed, 0, run.lane_zero_episode_index),
                "scale lane-zero episode seed mismatch",
            )?;
        }
        let first_run = self
            .runs
            .first()
            .ok_or_else(|| invalid("scale report has no runs"))?;
        let lane_zero_consistent = self.runs.iter().all(|run| {
            run.lane_zero_replay_digest == first_run.lane_zero_replay_digest
                && run.lane_zero_episode_index == first_run.lane_zero_episode_index
                && run.lane_zero_episode_seed == first_run.lane_zero_episode_seed
        });
        require(
            self.lane_zero_digest_consistent == lane_zero_consistent,
            "scale lane-zero consistency verdict mismatch",
        )?;
        let promotion_complete = self.requested_widths == SUPPORTED_BATCH_WIDTHS;
        require(
            self.promotion_widths_complete == promotion_complete,
            "scale promotion-width verdict mismatch",
        )?;
        self.runtime_contract.validate()?;
        let expected_passed = match self.evidence_class.as_str() {
            "contract_test" => {
                require(
                    self.backend_status == "test_only",
                    "contract-test scale evidence is not test-only",
                )?;
                validate_evidence_runtime("test_only", &self.runtime_contract, &self.runtime)?;
                lane_zero_consistent
            }
            "accelerator" => {
                require(
                    self.backend_status == "available",
                    "accelerator scale evidence is not available",
                )?;
                validate_evidence_runtime("available", &self.runtime_contract, &self.runtime)?;
                lane_zero_consistent && promotion_complete
            }
            _ => return Err(invalid("unknown scale evidence class")),
        };
        require(self.passed == expected_passed, "scale verdict mismatch")?;
        require(
            self.content_sha256 == self.recomputed_content_sha256()?,
            "scale content digest mismatch",
        )
    }

    /// Binds a valid report to its exact manifest, runtime contract, TaskSpec, and model bytes.
    pub fn validate_against(
        &self,
        manifest: &AcceleratorManifest,
        runtime_contract: &AcceleratorRuntimeContract,
        task_spec: &TaskSpec,
        model_bytes: &[u8],
    ) -> Result<(), AcceleratorContractError> {
        self.validate()?;
        manifest.validate()?;
        runtime_contract.validate()?;
        task_spec
            .validate()
            .map_err(|error| invalid(format!("bound TaskSpec is invalid: {error}")))?;
        require(
            self.adapter_id == manifest.id,
            "scale adapter differs from manifest",
        )?;
        require(
            self.precision == manifest.precision,
            "scale precision differs from manifest",
        )?;
        require(
            self.task_spec_schema == manifest.task_spec_schema,
            "scale TaskSpec schema differs from manifest",
        )?;
        require(
            self.requested_widths
                .iter()
                .all(|width| manifest.supported_batch_widths.contains(width)),
            "scale width is not advertised by manifest",
        )?;
        require(
            &self.runtime_contract == runtime_contract,
            "scale runtime contract differs from selected contract",
        )?;
        require(
            self.task_id == task_spec.task_id,
            "scale task id differs from TaskSpec",
        )?;
        require(
            self.task_spec_sha256 == task_spec_sha256(task_spec)?,
            "scale TaskSpec digest mismatch",
        )?;
        require(
            self.model_sha256 == normalized_model_sha256(model_bytes)?,
            "scale model digest mismatch",
        )
    }

    fn recomputed_content_sha256(&self) -> Result<String, AcceleratorContractError> {
        let mut value = serde_json::to_value(self)
            .map_err(|error| invalid(format!("serialize scale report: {error}")))?;
        value
            .as_object_mut()
            .ok_or_else(|| invalid("scale report did not serialize as an object"))?
            .remove("content_sha256");
        Ok(sha256(canonical_python_json(&value)?.as_bytes()))
    }
}

fn parse_raw_f64(value: &RawValue) -> Result<f64, AcceleratorContractError> {
    let parsed = value
        .get()
        .parse::<f64>()
        .map_err(|_| invalid("scale throughput is not an f64"))?;
    require(parsed.is_finite(), "scale throughput is not finite")?;
    Ok(parsed)
}

#[cfg(test)]
mod tests {
    use super::*;

    const MANIFEST: &str = include_str!("../../../adapters/mjx/accelerator.toml");
    const RUNTIME: &str = include_str!("../../../adapters/mjx/runtime.toml");
    const TASK: &str = include_str!("../../../adapters/mjx/fixtures/free-fall-task-spec-v1.json");
    const MODEL: &[u8] = include_bytes!("../../../adapters/mjx/fixtures/free-fall-v1.xml");
    const REPORT: &str = include_str!("../../../tests/golden/accelerators/scale-report-v1.json");

    fn contracts() -> (
        AcceleratorManifest,
        AcceleratorRuntimeContract,
        TaskSpec,
        AcceleratorScaleReport,
    ) {
        (
            toml::from_str(MANIFEST).unwrap(),
            toml::from_str(RUNTIME).unwrap(),
            serde_json::from_str(TASK).unwrap(),
            AcceleratorScaleReport::from_json_slice(REPORT.as_bytes()).unwrap(),
        )
    }

    #[test]
    fn selected_scale_report_recomputes_every_binding() {
        let (manifest, runtime, task, report) = contracts();
        report
            .validate_against(&manifest, &runtime, &task, MODEL)
            .unwrap();
    }

    #[test]
    fn timing_width_and_verdict_tampering_fail_closed() {
        let (_, _, _, mut report) = contracts();
        report.runs[0].transitions += 1;
        assert!(report.validate().is_err());
        let (_, _, _, mut report) = contracts();
        report.runs[0].throughput_transitions_s += 1.0;
        assert!(report.validate().is_err());
        let (_, _, _, mut report) = contracts();
        report.promotion_widths_complete = true;
        assert!(report.validate().is_err());
        let (_, _, _, mut report) = contracts();
        report.passed = false;
        assert!(report.validate().is_err());
    }

    #[test]
    fn lane_identity_task_model_and_digest_tampering_fail_closed() {
        let (_, _, _, mut report) = contracts();
        report.runs[1].lane_zero_episode_seed ^= 1;
        assert!(report.validate().is_err());
        let (_, _, _, mut report) = contracts();
        report.runs[1].lane_zero_episode_index += 1;
        assert!(report.validate().is_err());
        let (manifest, runtime, mut task, report) = contracts();
        task.task_id = "rne.physics.other.v1".into();
        assert!(report
            .validate_against(&manifest, &runtime, &task, MODEL)
            .is_err());
        let (manifest, runtime, task, report) = contracts();
        assert!(report
            .validate_against(&manifest, &runtime, &task, b"<mujoco/>")
            .is_err());
        let (_, _, _, mut report) = contracts();
        report.content_sha256.replace_range(..1, "0");
        assert!(report.validate().is_err());
    }
}
